cudax/include/cuda/experimental/__container/uninitialized_async_buffer.h

```diff
@@ -0,0 +1,218 @@
+  }
+
+  //! @brief Determines the properly aligned start of the buffer given the alignment and size of `T`
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr _Tp* __get_data() const noexcept
+  {
+    constexpr size_t __alignment = alignof(_Tp);
+    size_t __space               = __get_allocation_size(__count_);
+    void* __ptr                  = __buf_;
+    return reinterpret_cast<_Tp*>(_CUDA_VSTD::align(__alignment, __count_ * sizeof(_Tp), __ptr, __space));
+  }
+
+public:
+  using value_type = _Tp;
+  using reference  = _Tp&;
+  using pointer    = _Tp*;
+  using size_type  = size_t;
+
+  //! @brief Constructs a \c uninitialized_async_buffer, allocating sufficient storage for \p __count elements through
+  //! \p __mr
+  //! @param __mr The async memory resource to allocate the buffer with.
+  //! @param __stream The cuda stream used for stream ordered allocation.
+  //! @param __count The desired size of the buffer.
+  //! @note Depending on the alignment requirements of `T` the size of the underlying allocation might be larger
+  //! than `count * sizeof(T)`. Only allocates memory when \p __count > 0
+  uninitialized_async_buffer(
+    _CUDA_VMR::async_resource_ref<_Properties...> __mr, const ::cuda::stream_ref __stream, const size_t __count)
+      : __mr_(__mr)
+      , __stream_(__stream)
+      , __count_(__count)
+      , __buf_(__count_ == 0 ? nullptr : __mr_.allocate_async(__get_allocation_size(__count_), __stream_))
+  {}
+
+  uninitialized_async_buffer(const uninitialized_async_buffer&)            = delete;
+  uninitialized_async_buffer& operator=(const uninitialized_async_buffer&) = delete;
+
+  //! @brief Move construction
+  //! @param __other Another \c uninitialized_async_buffer
+  uninitialized_async_buffer(uninitialized_async_buffer&& __other) noexcept
+      : __mr_(__other.__mr_)
+      , __stream_(_CUDA_VSTD::exchange(__other.__stream_, {}))
+      , __count_(_CUDA_VSTD::exchange(__other.__count_, 0))
+      , __buf_(_CUDA_VSTD::exchange(__other.__buf_, nullptr))
+  {}
+
+  //! @brief Move assignment
+  //! @param __other Another \c uninitialized_async_buffer
+  uninitialized_async_buffer& operator=(uninitialized_async_buffer&& __other) noexcept
+  {
+    if (this == _CUDA_VSTD::addressof(__other))
+    {
+      return *this;
+    }
+
+    if (__buf_)
+    {
+      __mr_.deallocate_async(__buf_, __get_allocation_size(__count_), __stream_);
+    }
+    __mr_     = __other.__mr_;
+    __stream_ = _CUDA_VSTD::exchange(__other.__stream_, {});
+    __count_  = _CUDA_VSTD::exchange(__other.__count_, 0);
+    __buf_    = _CUDA_VSTD::exchange(__other.__buf_, nullptr);
+    return *this;
+  }
+
+  //! @brief Destroys an \c uninitialized_async_buffer deallocating the buffer
+  //! @warning The destructor does not destroy any objects that may or may not reside within the buffer. It is the users
+  //! responsibility to ensure that all objects within the buffer have been properly destroyed.
+  ~uninitialized_async_buffer()
```

Comment:
I mentioned this on your design doc, but want to mention it here too. In RMM we learned that requiring the user to set the correct CUDA device on destruction is onerous. [There is a very good explanation and example code that I encourage you to read here](https://github.com/rapidsai/rmm/issues/1342). In a nutshell, safely handling buffer constructor failures in multidevice code requires very hairy code. The upshot is that in `rmm::device_buffer` we [now store the current CUDA device](https://github.com/rapidsai/rmm/pull/1370) in the ctor and make it active in the dtor. 

See also https://github.com/rapidsai/rmm/issues/1527 and https://github.com/rapidsai/rmm/pull/1533.

Reply:
I believe we want to move towards a design, where we query the provided stream for the device and then use that @pciolkosz comments?

Reply:
I don't think cudaFree/cudaFreeAsync has any requirement on the current device. Maybe if NULL stream is used in the async version, but we plan to have no NULL stream anyway.